tempesta_fw/http_parser.c

```diff
@@ -4118,185 +4172,3193 @@ Req_Method_1CharStep: __attribute__((cold))
+{
+	TfwStr *str;
+	int r = CSTR_NEQ;
+	__FSM_DECLARE_VARS(req);
+
+	__FSM_START(parser->_i_st);
+
+	__FSM_STATE(Req_I_UriMarkStart) {
+		if (WARN_ON_ONCE(c != '/'))
+			__FSM_EXIT(CSTR_NEQ);
+
+		__msg_field_open(&req->mark, p);
+		/* Place initial slash into separate chunk. */
+		__FSM_H2_I_MOVE_LAMBDA_fixup_f(Req_I_UriMarkName, 1, &req->mark,
+		{
+			/*
+			 * The end of ':path' header has been met; thus, we can
+			 * just go out, and the parsed '/' will be fixed up in
+			 * the outside state after returning.
+			 */
+			TFW_STR_INIT(&req->mark);
+			return __data_off(p + 1);
+		}, 0);
+	}
+
+	__FSM_STATE(Req_I_UriMarkName) {
+		str = tfw_http_sess_mark_name();
+		H2_TRY_STR_2LAMBDA_fixup(str, &req->mark, {
+			parser->to_read = tfw_http_sess_mark_size();
+		}, {
+			__FSM_EXIT(CSTR_NEQ);
+		}, Req_I_UriMarkName, Req_I_UriMarkValue);
+		/*
+		 * In case of HTTP/2 processing we need not set @req->uri_path
+		 * here; instead, the value of ':path' pseudo-header in
+		 * @req->h_tbl (currently @parser->hdr) is used. If mark isn't
+		 * matched here, concatenate descriptors accumulated in
+		 * @req->mark with the descriptor of ':path' pseudo-header
+		 * (that is @parser->hdr) - the latter will be finished in the
+		 * 'Req_Path' state. Note, that if we are here, we must not be
+		 * postponed in the outside state after returning.
+		 */
+		if (tfw_strcat(req->pool, &parser->hdr, &req->mark))
+			__FSM_EXIT(CSTR_NEQ);
+
+		TFW_STR_INIT(&req->mark);
+		return __data_off(p);
+	}
+
+	__FSM_STATE(Req_I_UriMarkValue) {
+		__fsm_n = min_t(long, parser->to_read, __data_remain(p));
+		parser->to_read -= __fsm_n;
+		if (parser->to_read) {
+			if (fin)
+				__FSM_EXIT(CSTR_NEQ);
+			__msg_field_fixup_pos(&req->mark, p, __fsm_n);
+			__FSM_I_field_chunk_flags(&req->mark, TFW_STR_VALUE);
+			parser->_i_st = &&Req_I_UriMarkValue;
+			__FSM_EXIT(CSTR_POSTPONE);
+		}
+		parser->to_read = -1;
+		__msg_field_finish_pos(&req->mark, p, __fsm_n);
+		__FSM_I_field_chunk_flags(&req->mark, TFW_STR_VALUE);
+		return __data_off(p + __fsm_n);
+	}
+
+done:
+	return r;
+}
+STACK_FRAME_NON_STANDARD(__h2_req_parse_mark);
+
+int
+tfw_h2_parse_req_hdr(unsigned char *data, unsigned long len, TfwHttpReq *req,
+		     bool fin, bool value_stage)
+{
+	int ret = T_OK;
+	TfwMsgParseIter *it = &req->pit;
+	__FSM_DECLARE_VARS(req);
+
+	T_DBG("%s: len=%lu, data=%.*s%s, req=[%p]\n", __func__, len,
+	      min(500, (int)len), data, len > 500 ? "..." : "", req);
+
+	__FSM_START(parser->state);
+
+	/*
+	 * If next state is not defined here, that means the name has not been
+	 * parsed and is indexed, thus we should determine the state from the
+	 * header tag.
+	 */
+	__FSM_H2_REQ_NEXT_STATE(value_stage);
+
+	/* ----------------    (Pseudo-)headers name    ---------------- */
+
+	__FSM_STATE(RGen_Hdr, hot) {
+
+		tfw_http_msg_hdr_open(msg, p);
+
+		switch (c) {
+		case ':':
+			if (likely(__data_available(p, 7)
```

Comment:
Please use the same approach as for `c`-prefixed headers: check `__data_available(p, 5)` (the shortest pseudo header) and go to the slow path if we have less than 5 bytes available. Otherwise (likely) we should place all the pseudo headers from the sate in single switch statement by the 4-byte prefix cases (just as we do this for `c`-prefixed headers here).

I'd also propose to rewrite the whole state in this fashion: check for 5 available bytes and execut **single** and only switch on 4-bytes matching. This will give us about 20 cases which are solved in not more than 5 integer comparisons (and we'll be able to grow to 32 headers to keep the number of matches). This seems not a big deal, because we have only 2 other places with only 2 if statements each - not so many to optimize each case, however in total we'll replace the 4 if statements by only one additional comparison on the switch lookup table.

I'm not sure if the compiler properly moves `__data_remain(pos)` computation to a single place, so please use one of the `__FSM_DECLARE_VARS` variables (use `enum` if necessary to have a better name like `dremain`) and assign `data_remain()` value to it and use `dremain` in all the places instead of `_data_available()`. Probably it makes sense to introduce `data_available()` to work with the templarary variable.